kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,472 @@
+  public MongoWindowFlushManager init(
+      final int kafkaPartition
+  ) {
+    final WindowMetadataDoc metaDoc = metadata.findOneAndUpdate(
+        Filters.eq(WindowMetadataDoc.ID, kafkaPartition),
+        Updates.combine(
+            Updates.setOnInsert(WindowMetadataDoc.ID, kafkaPartition),
+            Updates.setOnInsert(WindowMetadataDoc.OFFSET, NO_COMMITTED_OFFSET),
+            Updates.setOnInsert(WindowMetadataDoc.STREAM_TIME, UNINITIALIZED_STREAM_TIME),
+            Updates.inc(WindowMetadataDoc.EPOCH, 1) // will set the value to 1 if it doesn't exist
+        ),
+        new FindOneAndUpdateOptions()
+            .upsert(true)
+            .returnDocument(ReturnDocument.AFTER)
+    );
+
+    if (metaDoc == null) {
+      throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
+    }
+
+    LOG.info("{}[{}] Retrieved initial metadata {}", name, kafkaPartition, metaDoc);
+
+    final var partitionSegments = new PartitionSegments(metaDoc.streamTime, metaDoc.epoch);
+
+    final var activeSegments = partitioner.activeSegments(kafkaPartition, metaDoc.streamTime);
+    if (activeSegments.isEmpty()) {
+      LOG.info("{}[{}] No active segments for initial streamTime {}",
+               name, kafkaPartition, metaDoc.streamTime);
+    } else {
+      for (final SegmentPartition segmentToCreate : activeSegments) {
+        // Most likely if we have active segments upon initialization, the corresponding
+        // collections and windowStart indices already exist, in which case #getCollection and
+        // #createIndex should be quick and won't rebuild either structure if one already exists
+        // TODO: optimize by using listCollectionNames to skip redundant collection/index creation
+        //  since we should almost always have an index for each collection for all active segments
+        partitionSegments.createSegment(segmentToCreate);
+      }
+
+      final long firstSegmentId = activeSegments.get(0).segmentId;
+      LOG.info("{}[{}] Initialized active segments in range {} - {}",
+               name, kafkaPartition, firstSegmentId, firstSegmentId + activeSegments.size());
+    }
+
+    kafkaPartitionToSegments.put(
+        kafkaPartition,
+        partitionSegments
+    );
+
+    return new MongoWindowFlushManager(
+        this,
+        (segment) -> windowsForSegmentPartition(kafkaPartition, segment),
+        metadata,
+        partitioner,
+        kafkaPartition
+    );
+  }
+
+  public String collectionNameForSegment(final SegmentPartition segmentPartition) {
+    return String.format(
+        "%d-%d",
+        segmentPartition.tablePartition, segmentPartition.segmentId
+    );
+  }
+
+  public MongoCollection<WindowDoc> windowsForSegmentPartition(
+      final int kafkaPartition,
+      final SegmentPartition segment
+  ) {
+    return kafkaPartitionToSegments.get(kafkaPartition).segmentWindows.get(segment);
+  }
+
+  public RemoteWriteResult<SegmentPartition> preCommit(
```

Comment:
it would be good to test the segment expiry/create logic as the stream time moves past the segment boundaries